src/RoomDirectoryModel.h

```diff
@@ -0,0 +1,64 @@
+// SPDX-FileCopyrightText: 2021 Nheko Contributors
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#pragma once
+
+#include <QAbstractListModel>
+#include "ui/FlatButton.h"
+
+#include <mtxclient/http/errors.hpp>
+#include "MatrixClient.h"
+
+#include <vector>
+
+class RoomDirectoryModel : public QAbstractListModel
+{
+public: 
+    explicit RoomDirectoryModel
+    (const std::string &s = "", QObject *parent = nullptr);
+
+    enum Roles {
+        // ??? TODO ???
+    };
+    QHash<int, QByteArray> roleNames() const override;
+
+    QVariant data(const QModelIndex &index, int role) const override;
+
+    inline int rowCount(const QModelIndex &parent = QModelIndex()) const override
+    {
+            (void)parent; // ???copied from UserModel - (why) do we need this???
+            return static_cast<int>(publicRoomsData_.size());
+    }
+
+    // the response's chunk field is a paginated array of PublicRoomsChunk data
+    // we only want to be able to view a few of these (eg. 5 rooms) at a time.
+    // Scrolling/Navigating to the next page should fetch the next set of rooms.
+    inline bool canFetchMore(const QModelIndex &parent) const override
```

Comment:
We want to be able to view more than 5 at a time, but we want to load them only when they get displayed the first time :3